design/aws-ecs/aws-ecs-service-addons.md

```diff
@@ -0,0 +1,108 @@
+- Users can enable any number of different combinations of features which were not accessible before. They can build the same load balanced services they could before, but now they can also enable App Mesh, X-Ray, Cloudwatch Agent, and FireLens on the service at the same time.
+- The code for each extension is its own self contained class. Rather than bloating the ECS constructs with logic that is specific to another service all of that logic can live in the extension itself.
+
+## Service Lifecycle
+
+The new `Service` construct differs a bit from other CDK L2+ constructs. Most CDK constructs are fully assembled in the constructor and are then immutable. The `Service` class is not like this. The constructor does nothing but initialize an empty list of extensions. Users add extensions to this list, to define their intention about what should be built later on. The service constructs do not actually exist at this point. The only thing that exists is a list of user specified intentions about what constructs they do want to exist.
+
+The service's constructs are instantiated once the service is added to an environment. At that point all the extensions are executed to build out the actual CDK resources. From this point on the service constructs are available and immutable.
+
+## Extension Implemention
+
+Each extension is a collection of functions which are executed by the `Service` class when the service is added to an environment. Functions are executed in the order that the extensions were added to the service. Some functions are mutating hooks which modify service or task definition properties prior to these constructs being created. Other functions are called after each construct has been created so that the extension can make use of the construct.
+
+The hooks for an extension are as follows:
+
+- `addHooks()` - This hook is called after all the extensions are added, but before anything else has been run. It gives each extension a chance to do some inspection of the overall service and see what other extensions have been added. Some addons may want to register hooks on the other addons to modify them. For example the Firelens addon wants to be able to modify the settings of the application container to route logs through Firelens.
+- `mutateTaskDefinitionProps()` - This is a mutating hook which is passed the `ecs.TaskDefinitionProps` for a TaskDefinition. This allows the extension to make modifications to the task definition, such as App Mesh extension modifying the proxy settings for the task.
+- `useTaskDefinition()` - After the task definition is created, this hook is passed the actual `TaskDefinition` construct that was created. This allows the extension to add containers to the task, modify the task definition's IAM role, etc.
```

Comment:
I also feel that `use` is a bit too fuzzy -- maybe `attachTo`? `connectTaskDefinition`?

Reply:
I don't think either `attach` or `connect` feels right to me. The purpose of the hook is to let the extension call methods on the created task definition, specifically methods like `TaskDefinition.addContainer()` but also to access `TaskDefinition.taskRole`

I didn't want to make it too specific like `addContainer` because not every extension will want to add a sidecar container. What do you think about switching to a model where the event is less "verb" and more event based, like the hook is called something like `taskDefinitionReady`